include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,646 @@
+        // q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta2 = exp(tau*(z*(-RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+
+        if (n%2) {
+            delta = delta2 - delta1;
+        } else {
+            delta = delta1 - delta2;
+        }
+
+        result += delta;
+        n++;
+    } while (abs(delta * sqrt(tau)) > eps || (abs(result * sqrt(tau)) > eps * eps && abs(delta/result) > eps));
+    
+    if (abs(result) < eps * eps)
+        return RealType(0);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta2tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    unsigned n = 0;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType /* q_n, */ delta, delta1, delta2, result = RealType(0);
+
+    do {
+        delta1 = exp(tau*(z*(RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+        // q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta2 = exp(tau*(z*(-RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+
+        delta = delta1 + delta2;
+
+        result += delta;
+        n++;
+    } while (abs(delta * sqrt(tau)) > eps || (abs(result * sqrt(tau)) > eps * eps && abs(delta/result) > eps));
+    
+    if (abs(result) < eps * eps)
+        return RealType(0);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta3tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    unsigned n = 1;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType /* q_n, */ delta, delta1, delta2, result = exp(-z*z*tau/constants::pi<RealType>());
+
+    do {
+        delta1 = exp(tau*(z*(RealType(2*n) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n)*RealType(n)));
+        // q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta2 = exp(tau*(z*(-RealType(2*n) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n)*RealType(n)));
+
+        delta = delta1 + delta2;
+
+        result += delta;
+        n++;
+    } while (abs(delta * sqrt(tau)) > eps || (abs(result * sqrt(tau)) > eps * eps && abs(delta/result) > eps));
+    
+    if (abs(result) < eps * eps)
+        return RealType(0);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta4tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    unsigned n = 1;
```

Comment:
I'd go with a 64 bit counter; I'll probably try to get a few million digits out of this and it's best if it doesn't wrap. Also, I'd venture the benchmarks will show there is no performance penalty.

Reply:
Well - `n` enters into the equation as `exp(-n*n)`. By the time `n` hits 1000 you'll have roughly 500,000 digits of accuracy.

For machine precision, `n` is lucky to hit 5!